packages/docusaurus-theme-classic/src/theme/DocItem/index.js

```diff
@@ -14,7 +14,75 @@ import DocPaginator from '@theme/DocPaginator';
 
 import styles from './styles.module.css';
 
+let headersCache = [];
+let tableOfContentLinksCache = [];
+let lastActiveTableOfContentLink;
+
+function getActiveHeader() {
+  const TOP_OFFSET = 80;
+  let index = 0;
+  let activeHeader;
+  while (index < headersCache.length && !activeHeader) {
+    const header = headersCache[index];
+    const {top} = header.getBoundingClientRect();
+    if (top > 0 && top <= TOP_OFFSET) {
+      activeHeader = header;
+    }
+    index += 1;
+  }
+  return activeHeader;
+}
+
+function highlightTableOfContentLink(tableOfContentLink) {
+  const cssClass = 'contents__link--active';
+  if (lastActiveTableOfContentLink) {
+    lastActiveTableOfContentLink.classList.remove(cssClass);
+  }
+  tableOfContentLink.classList.add(cssClass);
+  lastActiveTableOfContentLink = tableOfContentLink;
+}
+
+function setActiveTableOfContentLink() {
+  const activeHeader = getActiveHeader();
+  if (activeHeader) {
+    let index = 0;
+    let itemHighlighted = false;
+    while (index < tableOfContentLinksCache.length && !itemHighlighted) {
+      const tableOfContentLink = tableOfContentLinksCache[index];
+      if (activeHeader.innerText.indexOf(tableOfContentLink.innerText) > -1) {
+        highlightTableOfContentLink(tableOfContentLink);
+        itemHighlighted = true;
+      }
+      index += 1;
+    }
+  }
+}
+
+function addEventListeners() {
+  document.addEventListener('scroll', setActiveTableOfContentLink);
+  document.addEventListener('resize', setActiveTableOfContentLink);
+  document.addEventListener('hashchange', setActiveTableOfContentLink);
+}
+
+function removeEventListeners() {
+  document.removeEventListener('scroll', setActiveTableOfContentLink);
+  document.removeEventListener('resize', setActiveTableOfContentLink);
+  document.removeEventListener('hashchange', setActiveTableOfContentLink);
+}
+
+function componentLoaded() {
```

Comment:
The naming of this function can be better improved. You might not even need this function if we extract this whole feature out into a custom hook.

Reply:
I am trying to figure out where to place the hook since it might be shared among different themes.
I seen a couple of hooks here: docusaurus/packages/docusaurus/src/client/exports/
But I am not sure if this is the right place.
Can you point me in the right direction?
Thanks

Reply:
This is part of theme classic hook. So it should be under that package, not docusaurus (core). Hooks such as useDocusaurusContext is available in core because even if this theme is not used, its still valid for other site using other theme (eg: if someone authored theme-bootstrap)

Reply:
@endiliey thanks for your answer. Maybe I am not understanding but I feel like there is a contradiction between your answer and one of the changes @yangshun proposed.
He asked me to put the hook in a "shared" place so it can be used by different themes and now you are telling me it should be inside the theme classic structure.

Can you confirm me that it is going to be inside the custom theme and not in a "shared" place? 
In the case, it should be in a "shared" I feel like it should be in core but if not please indicate to me where should I put it.

Thanks!

Reply:
You can move it to core if you can make the hook very general. But i personally feel this is slightly too specific for theme classic because not all theme will want a right sidebar or active header. If they need to, its better for them to create their own IMO. We try to make core as small as possible

Reply:
Hmmm im feeling contradicted now too. The hook need to be very general though if placed in core. cc @yangshun for discussion first.

But if its too general, feels like it can be a npm package already 🤣

Reply:
Maybe we can have a `@docusaurus/utils` for random hook utils that don't belong in core but can be reused across packages. Anyway just put it in theme classic first :)

Reply:
Let's try this (tell me if you think is ok):
I going to push the changes when I finish adding this hook to core. if you guys think it is not generic enough to be there, then we can see how to improve the implementation or if it should be somewhere else.

sounds good?

Reply:
Yeah either way lets do small step
1. Refactor it into hook and put it in https://github.com/facebook/docusaurus/tree/master/packages/docusaurus-theme-classic/src/theme/hooks/<hookName>..js
You can import it as `@theme/hooks/<hookName>.js

The benefit of putting it in theme is that this is swizzle-able. Maybe some user are creative/ crazy enough to change the behavior a bit and instead make the current active header font size bigger by 3 times.

In core, you can't swizzle it

2. Make sure it works
3. Next step is something better talked in future